### PR TITLE
Add more features: sign In & place order

### PR DESCRIPTION
**Pull request by minhhaivu**
Just practice more with Git Bash

### CODE REVIEW

**Comment by minhhaivu** [comment]
Just practice

**Comment by minhhaivu** [comment]
Add sign in feature. Please review

**Comment by minhhaivu** [comment]
Add sign in feature

File: openweather/src/main/java/action/PlaceOrderAction.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        placeOrderPage = getPageInstance(PlaceOrderPage.class);
+    }
+
+    public PlaceOrderAction openMarketPlace() {
+        marketPlacePage.open();
+
+        return this;
+    }
+
+    public PlaceOrderAction selectProduct(String product) {
```
Comment (thuanlhv) [inline openweather/src/main/java/action/PlaceOrderAction.java]: "select product" is not a business action. I would suggest doing as below:

```
public PlaceOrderAction buyProduct(Product product) 
{
     // click Place Order
     // fill information provided in product
     // confirm order
}
```

File: openweather/src/main/java/action/PlaceOrderAction.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        marketPlacePage.selectProductToOrder(product);
+
+        return this;
+    }
+
+    public PlaceOrderAction addLocation(String searchStr, String searchBy) {
```
Comment (thuanlhv) [inline openweather/src/main/java/action/PlaceOrderAction.java]: Using string parameters is the most common problem of QAs, **DON'T DO THIS**.  It definitely leads to bad design. As I mentioned in the above comment, you define Product obj and it's done. Product obj should have all properties such as State, year, and so on.

File: openweather/src/main/java/action/PlaceOrderAction.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        return this;
+    }
+
+    public void orderHistoricalWeatherDataByState(String state, String year) {
+        placeOrderPage.selectState(state).selectYear(year).
+                submitOrderPlace();
+
+    }
+
+    public PlaceOrderAction filter(String[] weatherPara, String unit,
```
Comment (thuanlhv) [inline openweather/src/main/java/action/PlaceOrderAction.java]: I personally like `List<String>`, it's much easier to iterate the collection than Array.

File: openweather/src/main/java/action/SignInAction.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        signInPage = getPageInstance(SignInPage.class);
+        userHomePage = getPageInstance(UserHomePage.class);
+    }
+
+    public void signIn(String userEmail, String password) {
+        homePage.open().selectMenu("Sign in");
+        signInPage.enterSignInInfo(userEmail, password).submitSignIn();
+    }
+
+    public List<WebElement> getSuccessMsg() {
```
Comment (thuanlhv) [inline openweather/src/main/java/action/SignInAction.java]: this is not a business action, all those stuff should be handled in Validator classes

File: openweather/src/main/java/actor/SignInTester.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package actor;
+
+import action.SignInAction;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class SignInTester {
```
Comment (thuanlhv) [inline openweather/src/main/java/actor/SignInTester.java]: Don't break too small like this, ask yourself what is different between testing SignIn and Search Page?

File: openweather/src/main/java/pages/AbstractPage.java [hunk 6]
```diff
@@ -1,14 +1,41 @@
...
     public void close() {
         driver.close();
         driver.quit();
     }
+
+    public void sleep(long timeInSeconds) {
+        try {
+            Thread.sleep(timeInSeconds * 1000);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/AbstractPage.java]: Assume that you run on a single thread. This statement stops the test execution.

File: openweather/src/main/java/pages/AbstractPage.java [hunk 7]
```diff
@@ -1,14 +1,41 @@
...
+        }
+    }
+
+    public ExpectedCondition<Boolean> isPresent(By by) {
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/AbstractPage.java]: don't leave those things in AbstractPage, define `<Your Expected Condition.class>`

File: openweather/src/main/java/pages/HomePage.java [hunk 8]
```diff
@@ -38,8 +45,33 @@ public void submitSearch() {
         searchCityTxt.sendKeys(Keys.ENTER);
     }
 
+    public HomePage selectMenu(String menu) {
+        new WebDriverWait(driver, TIMEOUT_IN_SECONDS)
+                .until(ExpectedConditions.elementToBeClickable
+                        (By.xpath("//a[contains(text(),'" + menu + "')]")));
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/HomePage.java]: don't use selectors in page action method. Selector belongs to object repository which is used for mapping sector and web element id, try to separate them.

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    @FindBy(xpath = "//div[@class='google-map']//button[contains(text(),'Add location')]")
+    private WebElement addLocationBtn;
+
+    @FindBy(xpath = "//span[contains(text(),'Weather Parameters:')]")
+    private WebElement weatherParameterCbb;
+
+    @FindBy(xpath = "//div[@class='owm-check-box-group columns']")
+    private WebElement weatherParameterCkl;
+
+    @FindBy(xpath = "//div[@class='owm-check-box-group columns']/../../*[@class='icon-close']")
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java:52]: bad selector, use xpath axes https://www.w3schools.com/xml/xpath_axes.asp

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+    public PlaceOrderPage clickAddLocationButton() {
+        new WebDriverWait(driver, TIMEOUT_IN_SECONDS)
+                .until(ExpectedConditions.visibilityOf(addLocationBtn));
+        addLocationBtn.click();
+
+        return this;
+    }
+
+    public PlaceOrderPage selectTimePeriod(String fromYear, String fromMonth, String fromDate,
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java]: you should find a way that is more programming :(

File: openweather/src/main/java/pages/SignInPage.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    @FindBy(id = "user_password")
+    private WebElement passwordTxt;
+
+    @FindBy(xpath = "//div[@class='sign-form']/form[@id='new_user']/input[@name='commit']")
+    private WebElement submitBtn;
+
+    @FindBy(xpath = "//div[contains(text(),'Invalid Email or password.')]")
+    private List<WebElement> invalidSignMsg;
+
+    public SignInPage enterSignInInfo(String userEmail, String password) {
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/SignInPage.java:23]: in scope of SignInPage, you don't need to add SignIn to any method name.

File: openweather/src/main/java/test/search/CityDetailTest.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        List<SearchPage.CityInfo> expectedCities = cityList();
+        SearchTester tester = SearchTester.getInstance();
+        List<SearchPage.CityInfo> cityList = tester.searchAction.search("Ho Chi Minh");
+        Validator.checkResultDetail(cityList, expectedCities);
+        tester.searchAction.tearDown();
+    }
+
+    private static List<SearchPage.CityInfo> cityList() {
+        List<SearchPage.CityInfo> cities = new ArrayList<>();
+        cities.add(0, new SearchPage.CityInfo("Thanh pho Ho Chi Minh, VN", null));
```
Comment (thuanlhv) [inline openweather/src/main/java/test/search/CityDetailTest.java]: is an order of element important? if no don't specify it.

File: openweather/src/main/java/test/signin/Validator.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import org.openqa.selenium.WebElement;
+import org.testng.Assert;
+
+import java.util.List;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class Validator {
```
Comment (thuanlhv) [inline openweather/src/main/java/test/signin/Validator.java]: SignInValitor, you can easily find it by auto-complete code

File: openweather/src/main/java/action/PlaceOrderAction.java (hunk 2 above)
Comment (minhhaivu) [reply openweather/src/main/java/action/PlaceOrderAction.java]: Got it

File: openweather/pom.xml [hunk 14]
```diff
@@ -64,6 +64,12 @@
             <version>RELEASE</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
```
Comment (thuanlhv) [inline openweather/pom.xml:69]: why do we need this artifact?

File: openweather/src/main/java/element/CheckBox.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.NoArgsConstructor;
+import org.openqa.selenium.WebElement;
+
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
+public class CheckBox {
+    private static final String PRE_LABEL = "//label[contains(text(),'";
+    private static final String SUF_INPUT = "')]/input";
+    private static final String SUF_SPAN = "')]//span";
+
+    public static void select(WebElement checkBoxList, String checkBoxName) {
```
Comment (thuanlhv) [inline openweather/src/main/java/element/CheckBox.java]: we don't really need 2 methods, find a way to merge them into 1

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 16]
```diff
@@ -108,50 +105,57 @@ public PlaceOrderPage(WebDriver driver) {
...
-
-        return this;
-    }
-
-    public PlaceOrderPage clickAddLocationButton() {
-        new WebDriverWait(driver, TIMEOUT_IN_SECONDS)
-                .until(ExpectedConditions.visibilityOf(addLocationBtn));
-        addLocationBtn.click();
+    public PlaceOrderPage addLocation(Location location) {
+        if (location != null) {
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java:109]: I'm not sure what are use cases for the method 
- If I know that location is always null in some test cases, we just simply don't invoke the method
- If the location is null at runtime, I prefer to fail fast.

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 17]
```diff
@@ -108,50 +105,57 @@ public PlaceOrderPage(WebDriver driver) {
...
+            waitForPageLoaded();
+            searchTxt.click();
+            switch (location.searchBy) {
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java:112]: Don't use `switch case`, it's hard to maintain your code, simply write a private method.

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 18]
```diff
@@ -108,50 +105,57 @@ public PlaceOrderPage(WebDriver driver) {
...
-        fromDateInput.findElement(By.xpath(PRE_TEXT_1 + fromYear + SUF_TEXT_1)).click();
-        fromDateInput.findElement(By.xpath(PRE_TEXT_1 + fromMonth + SUF_TEXT_1)).click();
-        fromDateInput.findElement(By.xpath(PRE_TEXT_1 + fromDate + SUF_TEXT_1)).click();
-        toDateInput.click();
-        toDateInput.findElement(By.xpath(PRE_TEXT_1 + toYear + SUF_TEXT_1)).click();
-        toDateInput.findElement(By.xpath(PRE_TEXT_1 + toMonth + SUF_TEXT_1)).click();
-        toDateInput.findElement(By.xpath("//div[@class='to date-input']"
-                + PRE_TEXT_1 + toDate + SUF_TEXT_1)).click();
+    public PlaceOrderPage selectTimePeriod(Date fromDate, Date toDate) {
+        if (fromDate != null) {
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java:135]: same as the above comment, I prefer to fail fast if the value is unexpected at runtime

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 19]
```diff
@@ -108,50 +105,57 @@ public PlaceOrderPage(WebDriver driver) {
...
+            fromDateInput.click();
+            Calendar calender = Calendar.getInstance();
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java:137]: I suggest a method to handle it all 
```
// DateInput class

public void select(Date date) {
    // select year, month, ....
}
```

File: openweather/src/main/java/pages/PlaceOrderPage.java [hunk 20]
```diff
@@ -222,41 +216,49 @@ public void submitOrderPlace() {
...
     private void waitForPageLoaded() {
         new WebDriverWait(driver, TIMEOUT_IN_SECONDS).
-                until(isPresent(By.id("gmap")));
+                until(WaitForAction.isElementPresent(By.id("gmap")));
+    }
 
+    @Getter
+    @Setter
+    @NoArgsConstructor(access = AccessLevel.PUBLIC)
+    public static class Product {
```
Comment (thuanlhv) [inline openweather/src/main/java/pages/PlaceOrderPage.java:246]: How do we use 1 object to handle information for 3 different product types ?

**Comment by minhhaivu** [comment]
Change to new branch

**Pull request closed without merging**